Lab-2
Bhavesh Thadhani

In [65]:
import pandas as pd

# Loading dataset
df = pd.read_csv('../../practical_labs/datasets/Lab_2/Lab2_dataset.csv')

# Printing the dataset
print(df.head())


   Unnamed: 0 label                                               text   
0         605   ham  Subject: enron methanol ; meter # : 988291\nth...  \
1        2349   ham  Subject: hpl nom for january 9 , 2001\n( see a...   
2        3624   ham  Subject: neon retreat\nho ho ho , we ' re arou...   
3        4685  spam  Subject: photoshop , windows , office . cheap ...   
4        2030   ham  Subject: re : indian springs\nthis deal is to ...   

   label_num  
0          0  
1          0  
2          0  
3          1  
4          0  


In [66]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split

X = df['text']
y = df['label']

# Converting text data to numerical representation
vectorizer = CountVectorizer()
X = vectorizer.fit_transform(X)

# dividing the dataset into training and testing 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [67]:
from sklearn.svm import SVC
from sklearn.naive_bayes import MultinomialNB, GaussianNB
from sklearn.metrics import accuracy_score, classification_report

# Training SVC model
svc_model = SVC()
svc_model.fit(X_train, y_train)
y_pred_svc = svc_model.predict(X_test)
accuracy_svc = accuracy_score(y_test, y_pred_svc)
report_svc = classification_report(y_test, y_pred_svc)

# Training Multinomial Naive Bayes model
mnb_model = MultinomialNB()
mnb_model.fit(X_train, y_train)
y_pred_mnb = mnb_model.predict(X_test)
accuracy_mnb = accuracy_score(y_test, y_pred_mnb)
report_mnb = classification_report(y_test, y_pred_mnb)

# Training Gaussian Naive Bayes model
gnb_model = GaussianNB()
gnb_model.fit(X_train.toarray(), y_train)
y_pred_gnb = gnb_model.predict(X_test.toarray())
accuracy_gnb = accuracy_score(y_test, y_pred_gnb)
report_gnb = classification_report(y_test, y_pred_gnb)


In [68]:
# Comparing models
print("SVC Model Accuracy:", accuracy_svc)
print("SVC Model Classification Report:\n", report_svc)

print("\nMultinomial Naive Bayes Model Accuracy:", accuracy_mnb)
print("Multinomial Naive Bayes Model Classification Report:\n", report_mnb)

print("\nGaussian Naive Bayes Model Accuracy:", accuracy_gnb)
print("Gaussian Naive Bayes Model Classification Report:\n", report_gnb)

# Comment on the reasons behind the differences between models
# Include explanations based on model assumptions and dataset characteristics


SVC Model Accuracy: 0.9652173913043478
SVC Model Classification Report:
               precision    recall  f1-score   support

         ham       0.98      0.97      0.98       742
        spam       0.93      0.95      0.94       293

    accuracy                           0.97      1035
   macro avg       0.95      0.96      0.96      1035
weighted avg       0.97      0.97      0.97      1035


Multinomial Naive Bayes Model Accuracy: 0.978743961352657
Multinomial Naive Bayes Model Classification Report:
               precision    recall  f1-score   support

         ham       0.99      0.99      0.99       742
        spam       0.96      0.96      0.96       293

    accuracy                           0.98      1035
   macro avg       0.97      0.97      0.97      1035
weighted avg       0.98      0.98      0.98      1035


Gaussian Naive Bayes Model Accuracy: 0.9545893719806763
Gaussian Naive Bayes Model Classification Report:
               precision    recall  f1-score   suppor


I would say the SVC model correctly identified 96.52% of spam and non-spam emails. 
The Multinomial Naive Bayes model did even better, with a 97.87% accuracy. It's like a specialized tool for understanding and spotting spam.
The Gaussian Naive Bayes model, while still good, had a slightly lower accuracy of 95.46%. It's like a general-purpose tool that works well but not as perfectly as the specialized one.
The reasons for these differences may include how models make assumptions, the kind of data they work with, and how they are fine-tuned. The Multinomial Naive Bayes model, designed for text, performed the best because it's tailored for this type of task.

Overall, all models did well, but the specialized model was the best for this specific job.

In [69]:
import pandas as pd
import numpy as np

# Loading datset
df = pd.read_csv('../../practical_labs/datasets/Lab_2/AB_NYC_2019.csv')

def remove_outliers_zscore(dataframe, column, threshold=3):
    z_scores = np.abs((dataframe[column] - dataframe[column].mean()) / dataframe[column].std())
    return dataframe[z_scores <= threshold]

def remove_outliers_whiskers(dataframe, column):
    Q1 = dataframe[column].quantile(0.25)
    Q3 = dataframe[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    return dataframe[(dataframe[column] >= lower_bound) & (dataframe[column] <= upper_bound)]

cleaned_df_zscore = remove_outliers_zscore(df, 'price')

cleaned_df_whiskers = remove_outliers_whiskers(df, 'price')

print("Original DataFrame shape:", df.shape)
print("Cleaned DataFrame shape (Z-score approach):", cleaned_df_zscore.shape)
print("Cleaned DataFrame shape (Whiskers approach):", cleaned_df_whiskers.shape)



Original DataFrame shape: (48895, 16)
Cleaned DataFrame shape (Z-score approach): (48507, 16)
Cleaned DataFrame shape (Whiskers approach): (45923, 16)
